docarray/array/storage/redis/backend.py

```diff
@@ -0,0 +1,171 @@
+
+
+@dataclass
+class RedisConfig:
+    n_dim: int
+    host: str = field(default='localhost')
+    port: int = field(default=6379)
+    flush: bool = field(default=False)
+    update_schema: bool = field(default=True)
+    distance: str = field(default='COSINE')
+    redis_config: Dict[str, Any] = field(default_factory=dict)
+    index_text: bool = field(default=False)
+    tag_indices: List[str] = field(default_factory=list)
+    batch_size: int = field(default=64)
+    method: str = field(default='HNSW')
+    initial_cap: Optional[int] = None
+    ef_construction: Optional[int] = None
+    m: Optional[int] = None
+    ef_runtime: Optional[int] = None
+    block_size: Optional[int] = None
+    columns: Optional[List[Tuple[str, str]]] = None
+
+
+class BackendMixin(BaseBackendMixin):
+    """Provide necessary functions to enable this storage backend."""
+
+    TYPE_MAP = {
+        'str': TypeMap(type='text', converter=TextField),
+        'bytes': TypeMap(type='text', converter=TextField),
+        'int': TypeMap(type='integer', converter=NumericField),
+        'float': TypeMap(type='float', converter=NumericField),
+        'double': TypeMap(type='double', converter=NumericField),
+        'long': TypeMap(type='long', converter=NumericField),
+        # TODO add bool
+    }
+
+    def _init_storage(
+        self,
+        _docs: Optional['DocumentArraySourceType'] = None,
+        config: Optional[Union[RedisConfig, Dict]] = None,
+        **kwargs,
+    ):
+        if not config:
+            raise ValueError('Empty config is not allowed for Redis storage')
+        elif isinstance(config, dict):
+            config = dataclass_from_dict(RedisConfig, config)
+
+        if config.distance not in ['L2', 'IP', 'COSINE']:
+            raise ValueError(f'Distance metric {config.distance} not supported')
+        if config.method not in ['HNSW', 'FLAT']:
+            raise ValueError(f'Method {config.method} not supported')
+
+        if (
+            'decode_responses' in config.redis_config
+            and config.redis_config['decode_responses']
+        ):
+            config.redis_config['decode_responses'] = False
+
+        self._offset2id_key = 'offset2id'
+        self._config = config
+        self.n_dim = self._config.n_dim
+        self._config.columns = self._normalize_columns(self._config.columns)
+
+        self._client = self._build_client()
+        super()._init_storage()
+
+        if _docs is None:
+            return
+        elif isinstance(_docs, Iterable):
+            self.extend(_docs)
+        elif isinstance(_docs, Document):
+            self.append(_docs)
+
+    def _build_client(self):
+        client = Redis(
+            host=self._config.host,
+            port=self._config.port,
+            **self._config.redis_config,
+        )
+
+        if self._config.flush:
+            client.flushdb()
+
+        if self._config.update_schema:
+            if len(client.execute_command('FT._LIST')) > 0:
```

Comment:
what does FT._LIST means?

Reply:
FT._LIST returns a list of all existing indexes and this command is not wrapped in redis-py. Before dropping current index, we need to check if there is existing index, otherwise Redis will throw exception.